pyk/src/pyk/integration_tests/kompiled_test.py

```diff
@@ -0,0 +1,38 @@
+import shutil
+from abc import ABC
+from pathlib import Path
+from typing import Iterable, Optional
+from unittest import TestCase
+
+from ..ktool import KompileBackend, kompile
+
+
+class KompiledTest(TestCase, ABC):
+    KOMPILE_MAIN_FILE: str
+    KOMPILE_BACKEND: Optional[KompileBackend] = None
+    KOMPILE_OUTPUT_DIR: Optional[str] = None
+    KOMPILE_INCLUDE_DIRS: Iterable[str] = []
+    KOMPILE_EMIT_JSON = False
+
+    def setUp(self):
+        main_file = Path(self.KOMPILE_MAIN_FILE)
+        self.assertTrue(main_file.is_file())
+        self.assertEqual(main_file.suffix, '.k')
+
+        output_dir = Path(self.KOMPILE_OUTPUT_DIR) if self.KOMPILE_OUTPUT_DIR else None
+
+        include_dirs = [Path(include_dir) for include_dir in self.KOMPILE_INCLUDE_DIRS]
+        self.assertTrue(all(include_dir.is_dir() for include_dir in include_dirs))
+
+        self.kompiled_dir = kompile(
```

Comment:
Do you think it's beneficial to just make a `Kompile` class now, so that downstream we're already creating `Kompile` objects and calling `_.kompile()` on them, and won't have to make that change later?

I guess down the road, we will have things like:

```
kevm = Kompile(...)
....
kevm.init('before-lemmas', 'MODULE-NAME', 'llvm)                  // Actually calls kompile
...
kevm.add_lemmas_to_module(lemma_list, 'INTERNAL-MODULE')
...
kevm.init('after-lemmas', 'ANOTHER-MODULE', 'haskell')
```

Maybe this isn't a good pattern to use for some reason?

What do you think?

Reply:
I agree that we need a smart kompiler that only invokes the `kompile` command when necessary.

But maybe we don't actually need smart methods like `add_lemmas_to_module`. After all, we're going to read the main definition from disk, so perhaps we only need the following features.

* Collect transitive dependencies for a `KDefinition`.
* Compare timestamps like `make`.

This method of course doesn't have much domain-specific knowledge built in, so for example adding a simplification rule to a dependency triggers build. But projects should be organized in a way that these artifacts do not mix anyways, so I do not see this as a big issue.